networkx/algorithms/centrality/domirank.py

```diff
@@ -0,0 +1,342 @@
+    if patience <= 0:
+        raise nx.NetworkXAlgorithmError("it is mandatory that `patience > 0`.")
+    if dt <= 0 or dt >= 1:
+        raise nx.NetworkXAlgorithmError(
+            "it is mandatory that dt is bounded such that: `0 < dt < 1`."
+        )
+    if epsilon <= 0:
+        raise nx.NetworkXAlgorithmError(
+            "it is mandatory that `epsilon > 0` and recommended that `epsilon = 1e-5`."
+        )
+    GAdj = nx.to_scipy_sparse_array(G)  # convert to scipy sparse csr array
+
+    # Here we create a warning (I couldn't find a networkxwarning, only exceptions and erros), that suggests to use the iterative formulation of DomiRank rather than the analytical form.
+    if GAdj.shape[0] > 5000 and comp_method == "analytical":
+        import warnings
+
+        warnings.warn(
+            "The system is large!!! Consider using `comp_method = `iterative`` function argument for reduced computational time cost."
+        )
+    # Here we create another warning for alpha being supercharged
+    if alpha > 1:
+        if comp_method == "iterative":
+            raise nx.NetworkXUnfeasible(
+                "supercharging the competition parameter (`alpha > 1`) requires the `comp_method = `analytical`` flag."
+            )
+        else:
+            import warnings
+
+            warnings.warn(
+                "You are supercharging the competition in the system by having `alpha > 1`, which is only permitted for the analytical solution!"
+            )
+    if alpha < 0:
+        raise nx.NetworkXUnfeasible(
+            "the competition parameter alpha must be positive - `alpha > 0`."
+        )
+
+    # Here we renormalize alpha with the smallest eigenvalue (most negative eigenvalue) by calling the "hidden" function _find_smallest_eigenvalue()
+    # Note, this function always uses the iterative definition
+    sigma = np.abs(
+        alpha
+        / _find_smallest_eigenvalue(
+            GAdj,
+            max_depth=max_iter // 5,
+            dt=dt,
+            epsilon=epsilon,
+            max_iter=max_iter,
+            patience=patience,
+        )
+    )
+    match comp_method:
+        case "analytical":
+            converged = None
+            psi = sp.sparse.linalg.spsolve(
+                sigma * GAdj + sp.sparse.identity(GAdj.shape[0]),
+                sigma * GAdj.sum(axis=-1),
+            )
+        case "iterative":
+            psi, converged = _domirank_iterative(
+                GAdj,
+                sigma=sigma,
+                dt=dt,
+                epsilon=epsilon,
+                max_iter=max_iter,
+                patience=patience,
+            )
+        case _:
+            raise nx.NetworkXUnfeasible(
+                "The comp_method must match either `analytical` or the default `iterative` method."
+            )
+    psi = dict(zip(G, (psi).tolist()))
+    return psi, sigma, converged
+
+
+##### THE FUNCTIONS HEREUNDER SHOULD BE HIDDEN FUNCTIONS #####
```

Comment:
This comment can be deleted, as the functions names starting with `_` are not exposed outside of this file.